### ISSUE TITLE
abbr won't get triggered after a pipe and global option is not the solution

### ISSUE THREAD

**Issue by thingsiplay** [olets/zsh-abbr#53]
I am using zsh-abbr for a while now and have an issue. And I just updated it from git directly; also `zsh --version`: "zsh 5.8 (x86_64-pc-linux-gnu)". 

For me the abbreviations won't triggered or expanded after a pipe `|`. Like in this case: `x` expands to `chmod +x`, but `chmod +x | x` does not. This can be solved with a global option `-g`. The only problem is, that the abbr is really global and will even get triggered as an argument, like there: `grep x file` to `grep chmod +x file`. I would like to have an option that is less global and only affects as a command name, even after a pipe.

**Comment by olets** [comment]
Thanks for submitting the idea @thingsiplay! It makes sense. That's how zsh aliases work

```shell
% alias hello="echo world"
% echo hello; hello
hello
world
```

and how fish abbreviations work

```shell
> abbr --add w="echo world"
> echo hello; w
hello
world
``` 

I'll consider it but I'm not sure if it will be feasible. A naive version would use command abbreviations at the start of the line, and after at least `&&`, `||`, `;`, `|`, `[`, and `[[`; and `test` when that follows or word boundary or one of the delimiters I just listed. But there are probably important edge cases that would take truly parsing the command, and that's not in this project's scope.

I don't plan to start new abbr features soon. (I am getting close to launching some abbr updates though!) In the meantime you can create a global abbreviation and opt of using it in certain circumstances

```shell
% abbr -g x="chmod +x"
% grep chmod +x[ctrl space] file
```

or choose an abbreviation that doesn't conflict with things you type:

```shell
% abbr -g cx="chmod +x"
```

**Comment by thingsiplay** [comment]
I thought it wouldn't be easy, but better report than not to. I didn't know about the manual opt out, but this is still not the solution with a lot of abbreviations. Because I would need to make them all global in order to be able to use in pipes, at least those that make sense.

**Comment by olets** [comment]
> "[ctrl space]" as opposite if global option is not set

Interesting idea. That would not be as hard as parsing for command beginnings. I'll consider it. But I'm leaning towards not adding it— feels like a niche need that can be avoided with careful abbreviation naming. I'll leave this open though.

**Comment by thingsiplay** [comment]
Thanks for considering it, it's just an idea and I get why you won't add or change every niche idea. But in my personal opinion (I speak for myself) don't see this as such niche problem. This plugin itself is relatively niche, so you won't get too many reports on that. Maybe people are not aware that it's a problem, if they use the `-g` option and the command gets triggered at any point in the arguments. The first time it happened to me, I had weird results and didn't notice the expansion at all! It is not the pipe only, also when using `&&` or `;` in example are affected.

Careful naming won't change the problem, it will make it less likely to happen. But if I start making complex names, then the abbreviations lose their meaning. There is already a system how I name my abbreviations. My best workaround to ignore abbr when using global option currently is to put the arguments into quotation marks.

**Comment by olets** [comment]
Marked "wontfix" because I don't anticipate working on it, and "help wanted" because I'd consider contributions for it.

Anyone intending to take this on: please discuss here first.

**Comment by olets** [comment]
I don't plan to work on this (`wontfix`) but I'll consider contributions (`help wanted`).

Anyone intending to take this on: please discuss here first.

**Comment by olets** [comment]
[v6.4.0](https://github.com/olets/zsh-abbr/releases/tag/v6.4.0) adds experimental support for this!

- Read the full write-up at https://github.com/olets/zsh-abbr/pull/197

tl;dr yolo

```shell
# in .zshrc above where zsh-abbr is loaded

ABBR_EXPERIMENTAL_COMMAND_POSITION_REGULAR_ABBREVIATIONS=1
```

**Comment by thingsiplay** [comment]
Hi sorry for the late reply here, after years not using ZSH, I came back to ZSH with this plugin. Just encountered this issue again and to my surprise you are working on it! Thanks for this.

Right now I enabled the experimental global option and with my first few tests, it works as "advertised". For now, I need to use it before sharing my experience in the linked write up. But one question: Are there plans on auto expanding an abbr inside an abbr? Currently this doesn't even work with the global option (with or without). I'm not entirely sure if this is a different issue or needs to be addressed by this implementation:

```zsh
export ABBR_EXPERIMENTAL_COMMAND_POSITION_REGULAR_ABBREVIATIONS=2
```

```zsh
abbr "nvimupdate"="nvim -c \"lua require('lazy').update()\""
abbr -g "up"="echo hello ; nvimupdate"
```

or better yet:

```zsh
abbr "nvim update"="nvim -c \"lua require('lazy').update()\""
abbr -g "up"="echo hello ; nvim update"
```

The `nvimupdate` or `nvim update` works when typing it myself in the terminal, but they are not auto expanded when inside another abbreviation.

**Comment by olets** [comment]
Would either making `nvim update` an alias, or replacing

```
abbr -g "up"="echo hello ; nvim update"
```

with

```
abbr -g "up"="echo hello ; nvim -c \"lua require('lazy').update()\""
```

be a solution for you?

**Comment by thingsiplay** [comment]
I didn't think of using alias. Which works functional wise perfectly fine. But its not as good as first class support, because user-abbreviations does not evaluate alias (so file is no longer self contained) and history would not have expanded command. Overall this is not a big deal or anything like that. I was just reporting a potential issue. I will use the expanded command in the abbr like this:

```zsh
abbr -g "up"="echo hello ; nvim -c \"lua require('lazy').update()\""
```

Maybe the [documentation](https://zsh-abbr.olets.dev/configuration-variables.html) should have the information that an abbr is not evaluated inside another abbr when enabling `ABBR_EXPERIMENTAL_COMMAND_POSITION_REGULAR_ABBREVIATIONS`.
